HPCC-17280 Embedded R code uses persistent, shared environment

richardkchapman opened the pull request:
<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

## Testing:
See changes in regression suite
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

richardkchapman committed:
HPCC-17280 Embedded R code uses persistent, shared environment

Add options (as in Python) to control lifetime of the environment used by
embedded R code.

Note that the default behaviour (which was previously incorrect) will be
different - this should be mentioned in the RedBook.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman commented:
@johnholt FYI

richardkchapman commented:
@ghalliday Please review

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: 7114fef763e8a77587f2b0be6119a58fc27bb981
Build: failed
Number of Error(s): 5
/mnt/disk1/home/vamosax/smoketest/PR-9750/HPCC-Platform/plugins/Rembed/Rembed.cpp:182:86: error: invalid user-defined conversion from \xe2\x80\x98Rcpp::Environment_Impl<Rcpp::PreserveStorage>\xe2\x80\x99 to \xe2\x80\x98int\xe2\x80\x99 [-fpermissive]
/mnt/disk1/home/vamosax/smoketest/PR-9750/HPCC-Platform/plugins/Rembed/Rembed.cpp:693:86: error: invalid user-defined conversion from \xe2\x80\x98Rcpp::Environment_Impl<Rcpp::PreserveStorage>\xe2\x80\x99 to \xe2\x80\x98int\xe2\x80\x99 [-fpermissive]
make[2]: *** [plugins/Rembed/CMakeFiles/Rembed.dir/Rembed.cpp.o] Error 1
make[1]: *** [plugins/Rembed/CMakeFiles/Rembed.dir/all] Error 2
make: *** [all] Error 2
Get ECLWatch build result

richardkchapman commented:
Hmmm - I was compiling with current latest versions of Rcpp/RInside - not sure if that is why it works for me?

richardkchapman committed:
HPCC-17280 Embedded R code uses persistent, shared environment

Made a change to work around the fact that the two-parameter form of _new_env
does not appear to be present in older versions of Rcpp.

We could take the view that updating the version of Rcpp on the build machines
is preferable in whcih case this commit should be removed.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: d04e357f6fa8f10ab19e479e3aecc62b4ba5501a
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 91.702 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |16 |3 |0 |0 |



Regression test result:

| setup | hthor | thor | roxie |
| ----- | ----- | ----- | ---- |
| roxie_echo | Pass | Pass | Pass |
| roxie_keepwhitespace | Pass | Pass | Pass |
| setup | Pass | Pass | Pass |
| setup_fetch | Pass | Pass | Pass |
| setupdict | Pass | Pass | Pass |
| setupsearchindex | Pass | Pass | Pass |
| setupsq | Pass | Pass | Pass |
| setupwordindex | Pass | Pass | Pass |
| setupxml | Pass | Pass | Pass |
 

| test | hthor | thor | roxie |
| ----- | ----- | ----- | ---- |
| embedR | Pass | Pass | Pass |
| modelingWithR | Pass | Fail | Pass |
 
```xml
modelingWithR on thor failed with:
<Exception><Source>eclagent</Source><Message>System error: 0: Graph graph3[7], workunitwrite[9]: SLAVE #1 [10.241.40.11:20100]: Rembed: Rcpp error: object &apos;lmMdl&apos; not found,  - caused by (0, Rembed: Rcpp error: object &apos;lmMdl&apos; not found)</Message></Exception>

```
 
HPCC Stop: OK
HPCC Uninstall: OK

richardkchapman commented:
The failure on thor is highlighting a bad assumption in the new test code - I can mark it as nothor, or I can try to think about how to restructure it to avoid. @ghalliday Any ideas on that (I would prefer to make it work on thor too as it's intended to be an example as well as a test case..)

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: f7570c993f3acabaedf6dbf2baf2073e3db146fd
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 91.349 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |93 |93 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| setup | hthor | thor | roxie |
| ----- | ----- | ----- | ---- |
| roxie_echo | Pass | Pass | Pass |
| roxie_keepwhitespace | Pass | Pass | Pass |
| setup | Pass | Pass | Pass |
| setup_fetch | Pass | Pass | Pass |
| setupdict | Pass | Pass | Pass |
| setupsearchindex | Pass | Pass | Pass |
| setupsq | Pass | Pass | Pass |
| setupwordindex | Pass | Pass | Pass |
| setupxml | Pass | Pass | Pass |
 

| test | hthor | thor | roxie |
| ----- | ----- | ----- | ---- |
| embedR | Pass | Pass | Pass |
| modelingWithR | Pass | Fail | Pass |
 
```xml
modelingWithR on thor failed with:
<Exception><Source>eclagent</Source><Message>System error: 0: Graph graph3[7], workunitwrite[9]: SLAVE #1 [10.241.40.11:20100]: Rembed: Rcpp error: object &apos;lmMdl&apos; not found,  - caused by (0, Rembed: Rcpp error: object &apos;lmMdl&apos; not found)</Message></Exception>

```
 
HPCC Stop: OK
HPCC Uninstall: OK
